### PR TITLE
examples/gnrc_border_router: Add ethos and start_network.sh to README.md

### PR DESCRIPTION
**Pull request by kYc0o**
I had added it in a wrong branch, open again with correct commit message
Added documentation for ethos and automatic setup with the start_network.sh script on #4725

### CODE REVIEW

File: examples/gnrc_border_router/README.md [hunk 1]
```diff
@@ -60,3 +60,98 @@ for further help.
 [1] https://tools.ietf.org/html/rfc1055
 
 [2] https://github.com/contiki-os/contiki/blob/master/tools/tunslip.c
+
+
+# gnrc_border_router with ethos
+A new setup is possible thanks to \#4438(https://github.com/RIOT-OS/RIOT/pull/4438) on which only one serial interface is needed. This is done through the ethos (Ethernet Over Serial) driver that multiplexes serial data between ethernet packets and shell commands.
```
Comment (authmillenon) [inline examples/gnrc_border_router/README.md]: It's not typical to reference a merged PR in a documentation text.

Comment (authmillenon) [inline examples/gnrc_border_router/README.md]: Also, please try to keep the lines under 80 characters long.

File: examples/gnrc_border_router/README.md [hunk 2]
```diff
@@ -60,3 +60,98 @@ for further help.
...
+should compile the example and flash it to your board.
+
+Afterwards, you need to compile the ethos tool instead of tunslip.
+To do this, in `dist/tools/ethos` you'll find the tool, just compile it with:
+```bash
+$ make
+```
+
+## Usage
+Once you flashed the BR example on your board, you should intialize your IP settings by choosing a prefix for your 6lo network. Just type `make term` and setup your BR:
```
Comment (authmillenon) [inline examples/gnrc_border_router/README.md]: Shouldn't you rather use the `ethos` tool for that?

File: examples/gnrc_border_router/README.md (hunk 1 above)
Comment (kYc0o) [inline examples/gnrc_border_router/README.md]: Yes we'll wait until the PR is merged, then I'll change that part

File: examples/gnrc_border_router/README.md (hunk 2 above)
Comment (authmillenon) [inline examples/gnrc_border_router/README.md]: Ah the tun isn't initialized yet... Got it. (but maybe write below to kill this `make term` to prevent multiplexing issues over multiple terminals)

Comment (kYc0o) [inline examples/gnrc_border_router/README.md]: Oh yes! sorry, I'll change that

File: examples/gnrc_border_router/README.md [hunk 3]
```diff
@@ -60,3 +60,98 @@ for further help.
...
+## Setup
+Merge \#4725 to your current branch.
+Afterwards, compile `ethos` as explained previously.
+Then, you need to compile UHCP (micro host control protocol). This tool is found in `/dist/tools/uhcp`. So, as usual:
+```
+$ make clean all
+```
+
+At this point you should have these two programs compiled, necessary to compile your gnrc_border_router ** DO NOT COMPILE IT BEFORE ethos AND uhcp **.
+Then, proceed to compile and flash gnrc_border_router to your board as explained before.
```
Comment (authmillenon) [inline examples/gnrc_border_router/README.md]: Oops the formatting is a little broken here.

Comment (kYc0o) [inline examples/gnrc_border_router/README.md]: In the web editor looked well, I don't know why in the diff it is formatter that way...

Comment (authmillenon) [inline examples/gnrc_border_router/README.md]: Because it gets confused about the `** <text><linebreak><text> **` above

Comment (authmillenon) [inline examples/gnrc_border_router/README.md]: (just remove the spaces inside the boldness stars)

Comment (kYc0o) [inline examples/gnrc_border_router/README.md]: Ok thanks

File: examples/gnrc_border_router/README.md (hunk 1 above)
Comment (authmillenon) [inline examples/gnrc_border_router/README.md]: But #4438 is already merged

Comment (kYc0o) [inline examples/gnrc_border_router/README.md]: we are waiting for #4725

Comment (kYc0o) [inline examples/gnrc_border_router/README.md]: You mean I should remove the reference to the closed PR?

Comment (authmillenon) [inline examples/gnrc_border_router/README.md]: Yes. :-)

**Comment by gebart** [comment]
Can you change the prefix to anything in the ULA range (fc00::/7) instead of affe:: to use the RFC 4193 designated private use addresses

**Comment by authmillenon** [comment]
`2001:0000::/29 - 2001:01F8::/29` are also [viable for experimental use](https://tools.ietf.org/html/rfc4727#section-3.4.1)

File: examples/gnrc_border_router/README.md [hunk 4]
```diff
@@ -60,3 +60,99 @@ for further help.
 [1] https://tools.ietf.org/html/rfc1055
 
 [2] https://github.com/contiki-os/contiki/blob/master/tools/tunslip.c
+
+
+# gnrc_border_router with ethos
+A new setup is possible on which only one serial interface is needed. This is done through the ethos (Ethernet Over Serial) driver that multiplexes serial data between ethernet packets and shell commands.
```
Comment (authmillenon) [inline examples/gnrc_border_router/README.md]: ``s/ethos/`\0`/g``

Comment (authmillenon) [inline examples/gnrc_border_router/README.md]: And still too long lines

File: examples/gnrc_border_router/README.md [hunk 5]
```diff
@@ -60,3 +60,99 @@ for further help.
...
+
+## Requirements
+This functionality works only in Linux machines, Mac support will be added in the future (lack of `tap` interface), and Windows support should possible by setting up the needed `tap` interface with an external program (not tested).
```
Comment (authmillenon) [inline examples/gnrc_border_router/README.md]: `s/Mac/Mac OSX/`?

File: examples/gnrc_border_router/README.md [hunk 6]
```diff
@@ -60,3 +60,99 @@ for further help.
...
+
+## Setup
+Take care about the modification of the `Makefile`, since by default SLIP is enabled.
+In order to select ethos as the driver, you should modify some lines as following:
+```make
```
Comment (authmillenon) [inline examples/gnrc_border_router/README.md]: For better readability in `vim` or `cat` I would prefer a newline between prose text and code blocks.

File: examples/gnrc_border_router/README.md [hunk 7]
```diff
@@ -60,3 +60,99 @@ for further help.
...
+#CFLAGS += -DSLIP_UART=$(SLIP_UART)
+#CFLAGS += -DSLIP_BAUDRATE=$(SLIP_BAUDRATE)
+CFLAGS += '-DETHOS_UART=UART_DEV(0)' -DETHOS_BAUDRATE=115200 -DUSE_ETHOS_FOR_STDIO
```
Comment (authmillenon) [inline examples/gnrc_border_router/README.md]: I'm wondering if a diff wouldn't be better understandable, but I'm the git pro here :P

File: examples/gnrc_border_router/README.md [hunk 8]
```diff
@@ -60,3 +60,99 @@ for further help.
...
+Then deactivate the slip module and activate ethos
+```make
+# Include SLIP package for IP over Serial communication
+#USEMODULE += gnrc_slip
+USEMODULE += ethos gnrc_netdev2
+```
+This will allow to compile the BR firmware using ethos.
+Thus:
+```bash
+$ make clean all flash
```
Comment (authmillenon) [inline examples/gnrc_border_router/README.md]: `s/^$ //` if you syntax highlight as bash (otherwise `$` will be parsed as variable I think)

**Comment by gebart** [comment]
@authmillenon I interpret that paragraph that they mean more for routing experiments across sites for developing the IPv6 internet, not for local IoT experiments.

File: examples/gnrc_border_router/README.md [hunk 9]
```diff
@@ -60,3 +60,99 @@ for further help.
...
+> ifconfig 5 add affe::3/64
+> fibroute add :: via <link-local of tap> dev 6
+```
+
+Here we have choosen `affe` as the prefix. It should be disseminated to your other nodes on the radio range.
+
+Now, you should be able to ping your nodes using the global addres starting by your prefix, on our case `affe::`.
+
+# gnrc_border_router using automatic configuration (not merged yet)
+In \#4725 (https://github.com/RIOT-OS/RIOT/pull/4725) an effort to automate all the previous steps is done using 3 tools:
```
Comment (authmillenon) [inline examples/gnrc_border_router/README.md]: You can have proper links in markdown doing this

```markdown
[\#4725](https://github.com/RIOT-OS/RIOT/pull/4725)
```

(and it's not too disruptive for local readers either)

**Comment by authmillenon** [comment]
> @authmillenon I interpret that paragraph that they mean more for routing experiments across sites for developing the IPv6 internet, not for local IoT experiments.

Well the RFC also references `2001:db8::/32` for [documentation purposes](https://tools.ietf.org/html/rfc3849#section-2) which is exactly what we have here. :-)

**Comment by bapclenet** [comment]
Ping @authmillenon

**Comment by kYc0o** [comment]
Oops! That's me, I'll address the comments ASAP

File: examples/gnrc_border_router/README.md (hunk 4 above)
Comment (kYc0o) [inline examples/gnrc_border_router/README.md]: Why the long lines bother you? Do you think I should split into 2 paragraphs?

File: examples/gnrc_border_router/README.md (hunk 7 above)
Comment (kYc0o) [inline examples/gnrc_border_router/README.md]: Hahaha yeah, but you can teach me how to do it and we can try ;-)

File: examples/gnrc_border_router/README.md (hunk 8 above)
Comment (kYc0o) [inline examples/gnrc_border_router/README.md]: I didn't really catch `s/^$ //` ... :-(

Comment (OlegHahm) [inline examples/gnrc_border_router/README.md]: @authmillenon proposed to strip the `$` from the line.

**Comment by kYc0o** [comment]
I addressed most of the comments and changed the addresses for "documentation purposes" to `2001:db8::/32`.

**Comment by emmanuelsearch** [comment]
@kYc0o what's the status on this PR, on your side?
@aabadie do you see ways to improve this README based on your recent experience?

**Comment by kaspar030** [comment]
Current #4725 drops the slip mode completely. Should the readme reflect this?

**Comment by OlegHahm** [comment]
> Current #4725 drops the slip mode completely. Should the readme reflect this?

I think it might be nice to add a hint that you can use SLIP (or any Ethernet transceiver) as well.

**Comment by kYc0o** [comment]
On my side I think I addressed most of the comments, maybe a paragraph is apparently too large but I can't see how to split it or male it short. Otherwise we are still waiting for #4725

**Comment by kYc0o** [comment]
Looking at #4725 there are lots of changes so it seems this is very outdated. @kaspar030 can you confirm what is up-to-date and what's not?

**Comment by kYc0o** [comment]
As for my last tests, this guide is still valid, aiming to merge it once #4725 is merged too?

**Comment by kaspar030** [comment]
ACK after you shorten the lines to <= 80 characters. :)

**Comment by kYc0o** [comment]
Oh yes, I'm about to re-organise it because SLIP is not present anymore, I'll put first the config with the script and then with ethos, in case of someone wants to try the manual config with it. Some rephrase is needed too (#4725 is merged so we don't need to merge it locally anymore, etc.), but I think every step is here.

**Comment by kYc0o** [comment]
BTW, we'll get rid of SLIP?

**Comment by kaspar030** [comment]
> we'll get rid of SLIP?

I don't think so, as it's still perfectly working and standards compliant.

File: examples/gnrc_border_router/README.md [hunk 10]
```diff
@@ -1,4 +1,144 @@
-# gnrc_networking_border_router example
+# gnrc_border_router using automatic configuration
+A new setup is possible on which only one serial interface is needed. 
+This is done through the ethos (Ethernet Over Serial) driver and UHCP.
+Ethos multiplexes serial data to separate ethernet packets from shell commands.
+UHCP is in charge of configure the wireless interface prefix and routes on the BR.
```
Comment (aabadie) [inline examples/gnrc_border_router/README.md]: You should describe what UHCP means here (you did that a few lines below). Also `configure` => `configuring`

File: examples/gnrc_border_router/README.md [hunk 11]
```diff
@@ -1,4 +1,144 @@
...
+make clean all flash
+```
+
+## Usage
+Start the `start_network.sh` script by doing on `dist/tools/ethos`:
+
+```bash
+sudo sh start_network.sh /dev/ttyACMx tap0 2001:db8::/64
+```
+
```
Comment (aabadie) [inline examples/gnrc_border_router/README.md]: I think it would be great to mention here that one has access to the shell here. For instance, one can track the neighbour cache using `ncache`.

File: examples/gnrc_border_router/README.md [hunk 12]
```diff
@@ -1,4 +1,144 @@
...
+This will execute the needed commands to setup the network interface and configure the BR.
+Notice that this will also configure `2001:db8::/64` as a prefix.
+This prefix should be announced to other motes through the wireless interface.
+
+At this point you should be able to ping motes using the global address.
+You can do this through the RIOT console, for instance using `gnrc_networking`:
```
Comment (aabadie) [inline examples/gnrc_border_router/README.md]: To me, this is not clear. do you want to ping the motes behind the BR ? so why talking about gnrc_networking ? It's confusing.

File: examples/gnrc_border_router/README.md [hunk 13]
```diff
@@ -1,4 +1,144 @@
...
+```bash
+make clean all flash
+```
+
+On this RIOT BR, add some addresses and a route:
+
+```
+> ifconfig 6 add 2001:db8::2/48
+> ifconfig 5 add 2001:db8::3/64
+> fibroute add :: via <link-local of tap> dev 6
```
Comment (aabadie) [inline examples/gnrc_border_router/README.md]: You should tell what type of interface are 5 and 6 (wired/wireless). Looking at haw the BR is configured by `start_network.sh`, the BR wired interface has a link-local address `fe80:`

File: examples/gnrc_border_router/README.md [hunk 14]
```diff
@@ -33,21 +195,21 @@ the program by calling something like:
...
 ```
 and adding the SLIP interface to the neighbor cache (because Linux won't
 respond to neighbor solicitations on an interface without a link-layer address)
 by calling
 ```
-ncache add 6 affe::1
+ncache add 6 2001:db8::1
 ```
 After this you're basically done and should be able to ping between the border
 router and the outside world (assuming that the Linux host is properly
```
Comment (aabadie) [inline examples/gnrc_border_router/README.md]: To me there is a step missing here (and some more detail as well), because at this point the prefix is not propagated to the motes behind the BR (that was the case for me):
* detail that interface `6` is the wired interface and `5` is the wireless one, just for clarity
* configure the wireless interface with a global address using the same prefix:
```
ifconfig 5 add 2001:db8::3
```

**Comment by haukepetersen** [comment]
@aabadie: thanks for you input!

@kYc0o: could you adjust the README ASAP!? Would be good to merge this as soon as possible, as the border router app was merged and has a broken doc currently...

File: examples/gnrc_border_router/README.md [hunk 15]
```diff
@@ -1,4 +1,144 @@
-# gnrc_networking_border_router example
+# gnrc_border_router using automatic configuration
+A new setup is possible on which only one serial interface is needed.
```
Comment (emmanuelsearch) [inline examples/gnrc_border_router/README.md]: Since this is now the default config, how about rephrasing the first two sentences with something like "This setup uses a single serial interface, ethos (Ethernet Over Serial) and UHCP (micro Host Control Protocol).

File: examples/gnrc_border_router/README.md [hunk 16]
```diff
@@ -1,4 +1,144 @@
...
+```make
+ifeq (,$(filter native,$(BOARD)))
+GNRC_NETIF_NUMOF := 2
+USEMODULE += ethos gnrc_netdev2
+CFLAGS += '-DETHOS_UART=UART_DEV(0)' -DETHOS_BAUDRATE=115200 -DUSE_ETHOS_FOR_STDIO
+FEATURES_REQUIRED += periph_uart
+endif
+# include UHCP client
+USEMODULE += gnrc_uhcpc
+```
```
Comment (emmanuelsearch) [inline examples/gnrc_border_router/README.md]: Actually, do we need this in the readme at this point?
It seems that this check is superfluous in this example at least.
Should we move this to another part of the readme, e.g. in the manual config?

File: examples/gnrc_border_router/README.md [hunk 17]
```diff
@@ -1,4 +1,144 @@
...
+> ping6 fd00:dead:beef::1
+```
+
+Thus far, you can communicate using 6lowPAN and IPv6 between motes and your PC.
+
+## Issues
+A common issue is that after ~15min global addresses on the BR Neighbour Cache dissappear.
+We are investigating why it happens and we'll propose a solution soon.
+If you experience it, just reboot your nodes and they will come again on-line.
```
Comment (emmanuelsearch) [inline examples/gnrc_border_router/README.md]: reference the open issue on this here. I think this is #5122 right?

File: examples/gnrc_border_router/README.md (hunk 12 above)
Comment (kYc0o) [inline examples/gnrc_border_router/README.md]: I just want to point out that you can use global addresses to ping motes, either from linux or from other motes. I'll rephrase this to be more clear.

File: examples/gnrc_border_router/README.md (hunk 15 above)
Comment (kaspar030) [inline examples/gnrc_border_router/README.md]: > UHCP (micro Host Control Protocol)


micro host configuration protocol ;)

**Comment by kYc0o** [comment]
Comments addressed. Sorry for the delay, I was sick this morning... Please check if all comments were correctly addressed.

File: examples/gnrc_border_router/README.md [hunk 18]
```diff
@@ -1,4 +1,155 @@
...
+and configure the BR.
+Notice that this will also configure `2001:db8::/64` as a prefix.
+This prefix should be announced to other motes through the wireless interface.
+
+As said previously, `ethos` allows to send IP packets and shell commands.
+This is done through the same serial interface.
+By typing `help` you will get the list of available shell commands.
+
+At this point you should be able to ping motes using the global address.
+For instance, if you use the `gnrc_networking` example, you can ping with:
```
Comment (emmanuelsearch) [inline examples/gnrc_border_router/README.md]: "example" => "example on the mote"
"ping" => "ping it from your machine"

File: examples/gnrc_border_router/README.md [hunk 19]
```diff
@@ -1,4 +1,155 @@
...
+
+```
+> ping6 2001:db8:0:1234:0:567:8:1
+```
+
+Just replace this address by your mote's address.
+Using `ifconfig` shows you the addresses of your mote.
```
Comment (emmanuelsearch) [inline examples/gnrc_border_router/README.md]: Using `ifconfig`=> Using `ifconfig` on the shell of your mote

File: examples/gnrc_border_router/README.md [hunk 20]
```diff
@@ -1,4 +1,155 @@
...
+
+The script also sets up a link-local address on your Linux network interface.
+It can be pinged from 6lo motes:
```
Comment (emmanuelsearch) [inline examples/gnrc_border_router/README.md]: link-local => ULA (Unique Local Address)
(quite different, because routable inside the administrative domain)

File: examples/gnrc_border_router/README.md [hunk 21]
```diff
@@ -1,4 +1,155 @@
...
+
+```
+> ping6 fd00:dead:beef::1
+```
+
+Thus far, you can communicate using 6lowPAN and IPv6 between motes and your PC.
```
Comment (emmanuelsearch) [inline examples/gnrc_border_router/README.md]: Thus, IPv6 communication with between your PC and your motes is enabled?

File: examples/gnrc_border_router/README.md [hunk 22]
```diff
@@ -1,4 +1,155 @@
...
+
+## Issues
+A common issue [#5122](https://github.com/RIOT-OS/RIOT/issues/5122)
```
Comment (emmanuelsearch) [inline examples/gnrc_border_router/README.md]: A known issue?

**Comment by emmanuelsearch** [comment]
Thanks. Added a few comments. ACK when addressed.

File: examples/gnrc_border_router/README.md (hunk 20 above)
Comment (emmanuelsearch) [inline examples/gnrc_border_router/README.md]: Maybe also add: check your ULA on your PC with `ifconfig` command

**Comment by kYc0o** [comment]
Thanks to you @emmanuelsearch.

File: examples/gnrc_border_router/README.md [hunk 23]
```diff
@@ -1,4 +1,159 @@
...
+This will execute the needed commands to setup a `tap` interface 
+and configure the BR.
+Notice that this will also configure `2001:db8::/64` as a prefix.
+This prefix should be announced to other motes through the wireless interface.
+
+As said previously, `ethos` allows to send IP packets and shell commands.
+This is done through the same serial interface.
+By typing `help` you will get the list of available shell commands.
+
+At this point you should be able to ping motes using the global address.
```
Comment (aabadie) [inline examples/gnrc_border_router/README.md]: using the global address => using their global address

File: examples/gnrc_border_router/README.md [hunk 24]
```diff
@@ -1,4 +1,159 @@
...
+For instance, if you use the `gnrc_networking` example on the mote, you can
```
Comment (aabadie) [inline examples/gnrc_border_router/README.md]: Maybe use a link to point to [gnrc_networking](https://github.com/RIOT-OS/RIOT/tree/master/examples/gnrc_networking) and give a real example: run ifconfig one mote and display the output highlighting the global address. Then the next command will be more explicit.

File: examples/gnrc_border_router/README.md [hunk 25]
```diff
@@ -1,4 +1,159 @@
...
+```
+> ping6 2001:db8:0:1234:0:567:8:1
+```
+
+Just replace this address by your mote's address.
+Using `ifconfig` on the shell of your mote shows you the addresses of your 
+mote.
+
+The script also sets up a ULA (Unique Local Address) address on your Linux 
+network interface.
```
Comment (aabadie) [inline examples/gnrc_border_router/README.md]: Be explicit on the interface (I can have 4 interfaces on my Linux for instance).

File: examples/gnrc_border_router/README.md [hunk 26]
```diff
@@ -1,4 +1,159 @@
...
+
+Thus far, IPv6 communication with between your PC and your motes is enabled.
+
+## Issues
+A known issue [#5122](https://github.com/RIOT-OS/RIOT/issues/5122) 
+is that after ~15min global addresses on the BR Neighbour Cache dissappear.
+We are investigating why it happens and we'll propose a solution soon.
+If you experience it, just reboot your nodes and they will come again on-line.
+
+# gnrc_border_router with manual config
```
Comment (aabadie) [inline examples/gnrc_border_router/README.md]: "Manual configuration of gnrc_border_router" ?

File: examples/gnrc_border_router/README.md (hunk 24 above)
Comment (kYc0o) [inline examples/gnrc_border_router/README.md]: > Maybe use a link to point to gnrc_networking and give a real example: run ifconfig one mote and display the output highlighting the global address.

Mmm cannot get the point here... What do you mean with a link to point to gnrc_networking?

Comment (aabadie) [inline examples/gnrc_border_router/README.md]: I just meant to use the GitHub syntax to create an hyperlink to the gnrc_networking page:
```
[gnrc_networking](https://github.com/RIOT-OS/RIOT/tree/master/examples/gnrc_networking)
```

Comment (kYc0o) [inline examples/gnrc_border_router/README.md]: Mmm ok I can do that

File: examples/gnrc_border_router/README.md (hunk 26 above)
Comment (kYc0o) [inline examples/gnrc_border_router/README.md]: I'd name it as it was originally, something like gnrc_border_router example with manual config, what do you think?

Comment (emmanuelsearch) [inline examples/gnrc_border_router/README.md]: I guess it's fine. However: let's merge that asap, and also put it in the release candidate branch so that people have the required documentation to test...

Comment (kYc0o) [inline examples/gnrc_border_router/README.md]: OK I'll push my last addressed comments ASAP and try to put everything in the RC.

**Comment by kYc0o** [comment]
Since there's already an ACK, I'll squash the commits and merge. Since it is a readme I don't think Murdock is required for this. I'll let some minutes before doing it, in case of someone has more comments.

**Comment by emmanuelsearch** [comment]
re-ACK. Merge at will (and back port to release branch asap).

**Comment by kYc0o** [comment]
Squashed and go!

**Comment by authmillenon** [comment]
Needs backport.
